tools/phasestack.py

```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/env python3
+
+import sys
+from getopt import gnu_getopt, GetoptError
+from collections import namedtuple
+
+from libgcopdd import Event, UserError, open_dump_entry
+
+PhaseStack = namedtuple('PhaseStack', 'size, data')
+
+####
+
+def load(event):
+	with open_dump_entry(event + '.phasestack') as file:
+		# Use `sys.intern` so that `viewers.common`
+		# functions can use `is` instead of `==`.
+		# TODO: remove this after removing
+		#       `viewers.common` imports.
+		data = list(map(sys.intern, map(str.strip, file)))
+	return PhaseStack(len(data), data)
+
+def save(ps, event):
+	with open_dump_entry(event + '.phasestack', 'wt') as file:
+		file.write('\n'.join(ps.data) + '\n')
+
+####
+
+def get_mapping(single, aggregated):
```

Comment:
What is a mapping? Maybe using a class to define it and explain it's features would make it clearer.

Reply:
Consider also you wanted to aggregate the matrices M<sub>X</sub>, M<sub>X₂</sub>, M<sub>X₃</sub>, M<sub>X₄</sub> (that are corresponding to above mentioned phasestacks X, X₂, X₃, X₄).

You cannot just sum the matrices (they have potentially different headers and even sizes). You have to:

1. aggregate their phasestacks (and get A B C D E F, as said above)
2. expand all matrices so that their rows and columns correspond to lines of their aggregated phasestacks
3. sum the mapped matrices

Step 1 is `phasestack.aggregate` - the approximation of the shortest common supersequence (an NP-complete problem). Step 3 is trivial.

I will focus on step 2:

You will need to insert empty rows and columns. Or equivalently re-number existing rows, not using some numbers. And this is exactly what mapping is - again. If the matrix had all its rows and columns physically present, you could just assign it a mapping:

| change this | A | C | F |
| :---: | :---: | :---: | :---: |
| **A** | 42:43:44 | 1:1:1 | 45:46:47 |
| **C** | 1:1:1 | 1:1:1 | 1:1:1 |
| **F** | 48:49:50 | 1:1:1 | 51:52:53 |

| into this | A | *B* | C | *D* | *E* | F |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **A** | 42:43:44 | *0:0:0* | 1:1:1 | *0:0:0* | *0:0:0* | 45:46:47 |
| ***B*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| **C** | 1:1:1 | *0:0:0* | 1:1:1 | *0:0:0* | *0:0:0* | 1:1:1 |
| ***D*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| ***E*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| **F** | 48:49:50 | *0:0:0* | 1:1:1 | *0:0:0* | *0:0:0* | 51:52:53 |

The mapping is `[0, 2, 5]` - the one that you have seen above coming out of `get_mapping(X, Y)`.

But the matrix can be already missing rows and columns. `depmat.change_mapping` deals with this. It can compose the two mappings:

| change this | A | *C* | F |
| :---: | :---: | :---: | :---: |
| **A** | 42:43:44 | *0:0:0* | 45:46:47 |
| ***C*** | *0:0:0* | *0:0:0* | *0:0:0* |
| **F** | 48:49:50 | *0:0:0* | 51:52:53 |

| into this | A | *B* | *C* | *D* | *E* | F |
| :---: | :---: | :---: | :---: | :---: | :---: | :---: |
| **A** | 42:43:44 | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | 45:46:47 |
| ***B*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| ***C*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| ***D*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| ***E*** | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* |
| **F** | 48:49:50 | *0:0:0* | *0:0:0* | *0:0:0* | *0:0:0* | 51:52:53 |

This resulting matrix has the internal representation of `DepMat(size=2, mapping=[0, 5], data=[42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53])`.

The last thing I should add is that when `depmat.mapping is None`, it means it has no empty rows/columns. It is the same as `range(depmat.size)`, but allows simpler implementation of `add_to`.